### DISCUSSION TITLE
Генератор случайных треков

### DISCUSSION THREAD

**Discussion by Bondian** [Chimildic/goofy#136] [Идея] [answered]
Возможна ли на практике реализация генератора случайных ID треков? Например, генератор выдающий случайные число-буквенные комбинации в таком же формате, как у Spotify, в сочетании с функцией, проверяющей, существуют ли треки с такими идентификаторами в реальности?
Или эта идея утопична и нужно лишь надеяться, что в Spotify когда-нибудь появится кнопка "случайный трек"?

**Comment by Chimildic** [comment, marked as answer]
Мы не знаем как формируется _id_. Сгенерировать сотню другую строчек не проблема. Но не хватит ресурсов проверить их на существование. К тому же это совершенно нерелевантная выборка. 

Не совсем понятно о какой случайности идет речь: 
- Любая рекомендация за вычетом истории прослушиваний подойдет по определение случайного трека. 
- Если о функции перемешивания в плейлисте. Выборка треков ограничена, при рекомендованных плейлистах - повторяющиеся треки это болезнь Spotify, но никак не сортировки. 

Для себя объясняю отсутствие случайности сложным процессом взаимодействия. Представляем, что функция перемешивания ничего не знает о контексте. Дали набор треков, перемешали, вернули. Но мы в свою очередь уже слушали этот набор треков. Получается, что независимые друг от друга перемешивания действительно случайны, но в нашем контексте возникает искажение. Одни треки слышим чаще чем другие из-за того, что не завершаем прослушивание предыдущего независимого перемешивания.

Собственно изначальная цель goofy и была в том, чтобы избегать повторов всеми доступными способами. Удалять историю прослушиваний, вытеснять по дате добавления. Сюда в последствии добавилась возможность удаления треков. То есть перемешать один раз на нашей стороне, сохранить и постепенно удалять из плейлиста прослушанное. Таким способом обеспечиваем полное прослушивание одного перемешивания.

Забавы ради можно поиграться и с нерелевантными случайными треками
```js
function findRandomTracks() {
  let tracks = Source.mineTracks({
    keyword: [generate(), generate(), generate(), generate(), generate()],
  });
  if (tracks.length == 0) {
    findRandomTracks();
    return;
  }
  Filter.dedupArtists(tracks);
  Playlist.saveWithReplace({
    name: 'Случайные треки из поиска',
    tracks: tracks,
    randomCover: 'update',
  })

  function generate() {
    return Math.random().toString(36).replace(/[^a-z]+/g, '').substr(0, 5);;
  }
}
```

**Reply by Bondian** [reply]
> Мы не знаем как формируется _id_. Сгенерировать сотню другую строчек не проблема. Но не хватит ресурсов проверить их на существование. К тому же это совершенно нерелевантная выборка.

Собственно, это меня и интересовало. Просто когда-то где-то вычитал, что в глобальной базе Spotify постоянно находится более 20% треков никем никогда не прослушанных. В таком случае, существование кнопки "случайный трек" для любителей всевозможной "экзотики" было бы большим подарком, да и вообще любопытным инструментом для экспериментов. По теории вероятности, каждый пятый трек, выдаваемый этой кнопкой, был бы из числа подобных реликтовых экземпляров. Я подумал, что если есть способ генерировать случайные идентификаторы для Spotify, например, по аналогии с [Hex Color Code генератором](https://www.random.org/colors/hex), то это могло бы воплотить в жизнь идею такой кнопки. Но раз на проверку существования таких случайно сгенерированных "псевдо-идентификаторов" не хватит никаких ресурсов, то и смысла в этой затее нет. Существует ресурс [Forgotify](https://forgotify.com), выдающий "редкие" треки. Но он жутко неудобный, и часто не работает, к тому же часто выдаёт треки с многотысячными прослушиваниями... 


> Забавы ради можно поиграться и с нерелевантными случайными треками

Прошу прощения, но моих знаний хватает только для манипуляций с готовыми шаблонами. Я не понял как "запустить" приведённую выше функцию, чтобы с ней поиграться...

**Reply by Chimildic** [reply]
Нет, никакого подарка не будет. Горы нерелевантного мусора. Повторюсь, с лимитами платформы, это крайне неэффективное занятие. 

Для малоизвестных исполнителей можно изучать [EveryNoise](https://4pda.to/forum/index.php?s=&showtopic=248440&view=findpost&p=99521788). Или применять к выборке треков фильтр [rangeTracks](https://chimildic.github.io/goofy/#/func?id=rangetracks) с низким значением `followers`.

Обновил код, добавив сохранение в плейлист.

**Reply by Bondian** [reply]
Спасибо!
